src/state.rs

```diff
@@ -167,15 +204,16 @@ impl NetworkState {
         A: Canon,
         R: Canon,
     {
-        let mut context = CallContext::new(self);
+        let mut context =
+            CallContext::new(&mut self.latest, self.modules.clone());
 
         let result =
             context.query(target, Query::from_canon(&query), gas_meter)?;
 
         result.cast().map_err(VMError::from_store_error)
     }
```

Comment:
I am not sure we have a mechanism in place which makes it impossible for a query to change state - I guess not, and for this change I think we probably have to make sure that it is not allowed.

Reply:
Yeah, we don't have such a mechanism in place. An easy solution would be to clone `latest` and discard it, but I would be happier with a solution that takes `&self` instead of `&mut self.

Reply:
Fixed this by cloning and making it clear that a `query` doesn't change the state by taking a `&self` instead of `&mut self`. Currently this clones a snapshot - would be nice to not have to do that.